accepted/2021/preview-features/preview-features.md

```diff
@@ -0,0 +1,418 @@
+features. This can be used for auditing and for detecting transitive
+dependencies ([discussed later](#api-analyzer)).
+
+### runtime.json
+
+Whether or not `EnablePreviewFeatures` was set to true for the application
+should be written into the `runtime.json` file such that the host/runtime can
+use it to make decisions. Alternatively, we could say this is encoded by the
+fact that the assembly the host considers the entry point has the
+`RequiresPreviewFeatures` assembly level attribute.
+
+### Compilation context
+
+When `EnablePreviewFeatures` is true, the `LangVersion` property should be set
+to `Preview`. This avoids customers having to turn on preview features for the
+language separately.
+
+In addition, the property `EnablePreviewFeatures` should be passed to the
+compilation context (akin to how `TargetFramework` is passed in today). An
+analyzer will use that to block use of APIs that are marked as preview as basing
+this off the language preview mode feels wrong.
+
+***Note:*** F# doesn't have this capability today.
+
+### API Analyzer
+
+We'll need an attribute to record whether or not the project was
+built with preview features turned on:
+
+```C#
+namespace System.Runtime.Versioning
+{
+    [AttributeUsage(AttributeTargets.Assembly |
+                    AttributeTargets.Module |
+                    AttributeTargets.Class |
+                    AttributeTargets.Struct |
+                    AttributeTargets.Enum |
+                    AttributeTargets.Constructor |
+                    AttributeTargets.Method |
+                    AttributeTargets.Property |
+                    AttributeTargets.Field |
+                    AttributeTargets.Event, AllowMultiple = true, Inherited = false)]
+    public sealed class RequiresPreviewFeaturesAttribute : Attribute
+    {
+        public PreviewAttribute();
+    }
+}
+```
+
+We'll ship a built-in analyzer that will report diagnostics when an API is being
+used that is marked `[RequiresPreviewFeatures]` but the consumer (member, type,
+module, assembly) isn't marked. The default severity is `error`.
+
+This analyzer will also validate that if any referenced assembly has applied
+`[RequiresPreviewFeatures]`, that the consuming assembly must also have this
+attribute applied.
+
+User code will typically only have this attribute applied at the assembly-level.
+However, the framework will typically only use this attribute on types and
+members without putting it on the assembly itself.
+
+***Note:*** Since F# doesn't support analyzers, we would have to add this
+capability to the compiler itself.
+
+## Q & A
+
+### Are preview features supported?
+
+No. The entire point of preview feature is to gather feedback, and make changes
+based on it. This means we can't support them:
+
+* We will generally not service previews even if they are included in LTS
+  releases. The compiler has done this successfully already. If we can't hold
+  this line, then this effort has failed.
```

Comment:
How much trouble would we go to, if any, to avoid breaking preview features when we service supported releases?

Reply:
I think that depends:

* If the feature only requires work in the runtime/library space, then we probably get that for free since LTS releases are in their own branch anyway. We'd have to go out of our way to change how a preview feature works there.
* However, if the preview feature requires work in the SDK (which includes the compiler) then consumers would be limited by which SDK they install. We can probably make that happen by avoiding making breaking changes in preview features in minor updates of a past major. We should probably still reserve the right to make breaking changes changes in minor updates of the current major so that we can make improvements to preview features we're currently still previewing.